src/window/building/depot.c

```diff
@@ -0,0 +1,441 @@
+}
+
+int handle_mouse_depot_select_source_destination(const mouse *m, building_info_context *c, int is_source)
+{
+    if (scrollbar_handle_mouse(&scrollbar, m, 1)) {
+        return 1;
+    }
+    setup_for_selected_depot(c);
+
+    int y_offset = window_building_get_vertical_offset(c, 28);
+    int index = 0, offset = scrollbar.scroll_position;
+    for (int i = 0; i < MAX_VISIBLE_ROWS; i++) {
+        depot_select_storage_buttons[i].x = 0;
+        depot_select_storage_buttons[i].y = 22 * i;
+        depot_select_storage_buttons[i].width = 16 * (c->width_blocks - 2) - 4 -
+            (scrollbar.max_scroll_position > 0 ? 39 : 0);
+        depot_select_storage_buttons[i].height = ROW_HEIGHT;
+        depot_select_storage_buttons[i].left_click_handler = button_none;
+        depot_select_storage_buttons[i].parameter1 = 0;
+        depot_select_storage_buttons[i].parameter2 = 0;
+    }
+
+
+    for (building *b = building_first_of_type(BUILDING_GRANARY); b; b = b->next_of_type) {
+        if (index >= MAX_VISIBLE_ROWS) {
+            break;
+        }
+        if (building_storage_resource_max_storable(b, data.target_resource_id) > 0) {
+            if (offset > 0) {
+                offset--;
+            } else {
+                depot_select_storage_buttons[index].left_click_handler = is_source ? set_order_source : set_order_destination;
+                depot_select_storage_buttons[index].parameter1 = data.depot_building_id;
+                depot_select_storage_buttons[index].parameter2 = b->id;
+                index++;
+            }
+        }
+    }
+    for (building *b = building_first_of_type(BUILDING_WAREHOUSE); b; b = b->next_of_type) {
+        if (index == MAX_VISIBLE_ROWS) {
+            break;
+        }
+        if (building_storage_resource_max_storable(b, data.target_resource_id) > 0) {
+            if (offset > 0) {
+                offset--;
+            } else {
+                depot_select_storage_buttons[index].left_click_handler = is_source ? set_order_source : set_order_destination;
+                depot_select_storage_buttons[index].parameter1 = data.depot_building_id;
+                depot_select_storage_buttons[index].parameter2 = b->id;
+                index++;
+            }
+        }
+    }
+
+    return generic_buttons_handle_mouse(m, c->x_offset + 18, y_offset + 46, depot_select_storage_buttons,
+        MAX_VISIBLE_ROWS, &data.storage_building_focus_button_id);
+}
+
+int window_building_handle_mouse_depot_select_source(const mouse *m, building_info_context *c)
+{
+    return handle_mouse_depot_select_source_destination(m, c, 1);
+}
+
+void window_building_draw_depot_select_destination(building_info_context *c)
+{
+    draw_order_source_destination_background(c, translation_for(TR_BUILDING_INFO_DEPOT_SELECT_DESTINATION_TITLE));
+}
+
+void window_building_draw_depot_select_destination_foreground(building_info_context *c)
+{
+    draw_depot_select_source_destination(c);
```

Comment:
I don't understand the need for function duplication.

There are two outside drawing functions, one for the source and the other for the destination. But they both end up calling the function that draws everything.

This entire thing should either be consolidated into one single function or the source and destination drawing codes should be separated.

Reply:
Done.